src/crimson/osd/pg.cc

```diff
@@ -850,23 +851,29 @@ PG::do_osd_ops_execute(
       } else if (m.has_flag(CEPH_OSD_FLAG_FULL_TRY)) {
         // they tried, they failed.
         logger().info(" full, replying to FULL_TRY op");
-        if (get_pgpool().info.has_flag(pg_pool_t::FLAG_FULL_QUOTA))
-          return interruptor::make_ready_future<OpsExecuter::rep_op_fut_tuple>(
-            seastar::now(),
-            OpsExecuter::osd_op_ierrorator::future<>(
-              crimson::ct_error::edquot::make()));
-        else
+        if (get_pgpool().info.has_flag(pg_pool_t::FLAG_FULL_QUOTA)) {
+          return (*error_func_ptr)(EDQUOT, rep_tid).then([] {
+            return interruptor::make_ready_future<OpsExecuter::rep_op_fut_tuple>(
+              seastar::now(),
+              OpsExecuter::osd_op_ierrorator::future<>(
+                crimson::ct_error::edquot::make()));
+          });
+        } else {
+          return (*error_func_ptr)(ENOSPC, rep_tid).then([] {
+            return interruptor::make_ready_future<OpsExecuter::rep_op_fut_tuple>(
+              seastar::now(),
+              OpsExecuter::osd_op_ierrorator::future<>(
+                crimson::ct_error::enospc::make()));
+          });
+        }
+      } else {
+        // drop request
+        return (*error_func_ptr)(EAGAIN, rep_tid).then([] {
```

Comment:
We shouldn't be calling submit_error_log in these cases.

submit_error_log records the result of an IO into the pg log so that we can return the same error code if the client resends the request.  This is *only relevant* for logical errors resulting from the target object state -- for example, EEXIST returned on an exclusive create -- because there is application logic built to rely on them.  In classic, the only such site is if the return value from do_osd_ops is negative (or the transaction is empty) -- see PrimaryLogPG::prepare_transaction, specifically where we set update_log_only to true.

We do not want to record space usage errors or errors specific to conditions on the primary OSD such as IO errors.  submit_error_log isn't a catch-all error path.